### PR TITLE
feat(recipe): support structured parameters (object/array) in recipe templates

### PR DESCRIPTION
**Pull request by jordigilh**
## Summary

Add `object` and `array` variants to `RecipeParameterInputType` and wire
structured parameter support end-to-end through the build pipeline, enabling:

- **Dot-notation access** for objects: `{{ signal.namespace }}`
- **Iteration** for arrays: `{% for item in findings %}{{ item.name }}{% endfor %}`
- **Conditionals** on structured fields: `{% if signal.severity == "critical" %}`

Callers pass JSON strings via the existing `Vec<(String, String)>` API; the
build pipeline detects `input_type: object`/`array`, parses the strings into
`serde_json::Value`, and routes through the structured MiniJinja renderer.
No API signature changes to `build_recipe_from_template` or its callers.

**Related discussion:** https://github.com/block/goose/discussions/8917

## Motivation

The current recipe template rendering accepts only `HashMap<String, String>`,
requiring callers to flatten structured data into individual string keys.
For recipes that process complex inputs (multi-field signals, arrays of
findings, enrichment metadata), this creates an impedance mismatch --
MiniJinja natively supports dot-notation, iteration, and conditionals on
structured data.

## Changes

| File | Change |
|------|--------|
| \`crates/goose/src/recipe/mod.rs\` | Add \`Object\` and \`Array\` to \`RecipeParameterInputType\` enum |
| \`crates/goose/src/recipe/template_recipe.rs\` | Add \`render_recipe_content_with_structured_params\` fn + 10 unit tests |
| \`crates/goose/src/recipe/build_recipe/mod.rs\` | Wire structured rendering into \`render_recipe_template\` + \`to_structured_params\` helper |
| \`crates/goose/src/recipe/build_recipe/tests.rs\` | 4 integration tests through \`build_recipe_from_template\` |
| \`documentation/docs/guides/recipes/recipe-reference.md\` | Document \`object\` and \`array\` input types |

### How it works

1. \`render_recipe_template\` checks if any recipe parameter has \`input_type: object\` or \`array\`
2. If yes, \`to_structured_params\` converts the string map to \`HashMap<String, serde_json::Value>\` -- only keys matching object/array params are parsed as JSON; all others remain as \`Value::String\`
3. The structured map is passed to \`render_recipe_content_with_structured_params\`, which feeds it directly to MiniJinja
4. If no structured params exist, the existing string path is used (zero overhead)

All callers (\`goose-cli\`, \`goose-server\`, \`summon\`, \`execute_commands\`) benefit automatically since the change is internal to the build pipeline.

### Scope and limitations

- **Desktop/CLI UX**: Recipe editor dropdown and CLI \`--params\` prompting degrade gracefully (text input fallback for unknown types) but don't yet provide JSON-aware input widgets -- these are future work for the respective teams
- **API surface**: \`PUT /sessions/{id}/user_recipe_values\` body remains \`HashMap<String, String>\` -- callers pass JSON-encoded strings for object/array params

### Security model

The new function uses the **same MiniJinja \`Environment\` setup** as the existing
\`render_recipe_content_with_params\`:
- Same \`UndefinedBehavior::Strict\`
- Same \`recipe_dir\`-scoped template loader for \`extends\`/\`include\`
- No additional filesystem access, eval, or code execution capabilities
- MiniJinja's expression model prevents arbitrary code execution regardless
  of parameter structure

## Test plan

- [x] \`cargo check -p goose --lib\` passes
- [x] \`cargo fmt -p goose -- --check\` clean
- [x] \`cargo clippy\` introduces no new warnings
- [x] 10 unit tests in \`template_recipe.rs\` (string compat, object dot-notation, conditionals, arrays, nesting, optional fields, mixed params, edge cases)
- [x] 4 serde round-trip tests in \`mod.rs\` (JSON object, JSON array, YAML mixed, enum round-trip)
- [x] 4 integration tests in \`build_recipe/tests.rs\` (object param, array param, mixed string+object, invalid JSON error)
- [x] Containerized test run on Fedora (all tests pass)

### CODE REVIEW

File: crates/goose/src/recipe/mod.rs [hunk 1]
```diff
@@ -181,6 +181,12 @@ pub enum RecipeParameterInputType {
     /// Cannot have default values to prevent importing sensitive user files.
     File,
     Select,
+    /// Structured object parameter passed as JSON.
+    /// Enables dot-notation access in templates: `{{ param.field }}`
+    Object,
+    /// Array parameter passed as a JSON array.
+    /// Enables iteration in templates: `{% for item in param %}`
+    Array,
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply crates/goose/src/recipe/mod.rs:189]: Thanks for calling this out. The latest push (2de4daa) addresses this: `render_recipe_template` now detects `input_type: object`/`array` and routes through `render_recipe_content_with_structured_params`, parsing the JSON string values into `serde_json::Value` before MiniJinja rendering. All existing callers benefit without API changes. Four integration tests through `build_recipe_from_template` cover this end-to-end (object, array, mixed, and invalid-JSON error cases).

File: documentation/docs/guides/recipes/recipe-reference.md [hunk 2]
```diff
@@ -350,14 +352,31 @@ parameters:
...
-prompt: "Process {{ max_files }} files in {{ output_format }} format. Debug: {{ enable_debug }}. Code:\n\n{{ source_code }}"
+  - key: signal
+    input_type: object
+    requirement: required
+    description: "Signal data with name and namespace fields"
+
+  - key: findings
+    input_type: array
+    requirement: optional
+    description: "List of diagnostic findings"
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply documentation/docs/guides/recipes/recipe-reference.md:364]: Good catch. Fixed in 2de4daa -- added `default: "[]"` to the `findings` example so it passes `validate_optional_parameters`.

File: crates/goose/src/recipe/build_recipe/mod.rs [hunk 3]
```diff
@@ -32,18 +35,72 @@ where
...
+        .iter()
+        .map(|(k, v)| {
+            let value = if structured_keys.contains(k.as_str()) {
+                serde_json::from_str(v).map_err(|e| {
+                    anyhow::anyhow!(
+                        "Parameter '{}' has input_type object/array but value is not valid JSON: {}",
+                        k,
+                        e
+                    )
+                })?
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply crates/goose/src/recipe/build_recipe/mod.rs]: Good call. Fixed in 83db531 -- `to_structured_params` now validates that `input_type: object` receives a JSON object and `input_type: array` receives a JSON array, returning a targeted error like `Parameter 'signal' has input_type object but received an array` instead of letting mismatches surface as confusing MiniJinja template errors. Two new tests cover both mismatch directions.

File: crates/goose/src/recipe/template_recipe.rs [hunk 4]
```diff
@@ -172,10 +208,14 @@ pub fn parse_recipe_content(
...
 
+    // Chainable allows attribute access on undefined values (e.g., {{ signal.name }})
+    // to return undefined instead of erroring. This is needed because structured
+    // parameters (object/array) use dot-notation in templates, and during this
+    // parsing phase the template is rendered with no parameters to discover variables.
     let (env, template_variables) = get_env_with_template_variables(
         &preprocessed_content,
         recipe_dir,
-        UndefinedBehavior::Lenient,
+        UndefinedBehavior::Chainable,
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply crates/goose/src/recipe/template_recipe.rs:210]: This is already handled by the validation layer. `validate_parameters_in_template` only normalizes dot-access for parameters with `input_type: object` or `array`. For a scalar parameter `name`, the template variable `name.field` is NOT normalized, so the validation correctly reports "missing: name.field, unnecessary: name", rejecting the recipe before it reaches the renderer.

File: crates/goose/src/recipe/build_recipe/mod.rs [hunk 5]
```diff
@@ -32,18 +35,101 @@ where
...
+        .iter()
+        .map(|(k, v)| {
+            let value = if let Some(input_type) = structured_keys.get(k.as_str()) {
+                let parsed: Value = serde_json::from_str(v).map_err(|e| {
+                    anyhow::anyhow!(
+                        "Parameter '{}' has input_type {} but value is not valid JSON: {}",
+                        k,
+                        input_type,
+                        e
+                    )
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply crates/goose/src/recipe/build_recipe/mod.rs]: Good catch. Fixed in 3e95363 -- `validate_optional_parameters` now checks that object/array parameter defaults are valid JSON matching their `input_type` (object must be `{}`, array must be `[]`). Invalid defaults are rejected during recipe validation instead of surfacing as late runtime errors.

File: crates/goose/src/recipe/validate_recipe.rs [hunk 6]
```diff
@@ -94,19 +94,43 @@ fn validate_parameters_in_template(
...
+        })
+        .map(|p| p.key.as_str())
+        .collect();
+
+    let normalized_template_vars: HashSet<String> = template_variables
+        .iter()
+        .map(|var| {
+            let root = var.split('.').next().unwrap_or(var);
+            if structured_keys.contains(root) {
+                root.to_string()
```
[the remark this answers is not in the document]
Comment (jordigilh) [reply crates/goose/src/recipe/validate_recipe.rs]: Good catch. Fixed in 3e95363 -- `extract_root_identifier` now handles both dot-notation (`signal.name`) and bracket notation (`findings[0].name`, `signal["namespace"]`) by finding the earliest `.` or `[` delimiter to extract the root identifier.

**Comment by jordigilh** [comment]
@lifeizhou-ap @baxen @DOsinga could I get a pair of eyes on this PR? Basically the PR adds a nested object as a parameter. It's additive so it doesn't cause regressions. 
Last week I created a [discussion](https://github.com/aaif-goose/goose/discussions/8917) with so far no comments. This is the PR that implements the change in the discussion.

Thank you for taking a look. 

/Jordi

**Review by DOsinga — COMMENTED** [review]
I've pushed a cleanup commit that:

- **Merges `main`** — the branch was 82 commits behind, causing stale diffs in `openapi.json` and `types.gen.ts` (version downgrade, removed fields, etc.). Those generated files now only contain the actual enum additions (`object`, `array`).
- **Removes the trivial serde roundtrip test** (`test_object_array_input_type_serde_roundtrip`) — this just tests that `#[derive(Serialize, Deserialize)]` works on a simple enum variant, which is guaranteed by the derive macro.
- **Removes the verbose doc comment** on the private one-liner `extract_root_identifier` — the function name and signature are self-documenting.
- **Removes the placeholder TODO** in `recipeFormSchema.ts` that referenced `https://github.com/aaif-goose/goose/issues/NEW`.

The merge also resolved a subtle regression in the original PR's refactored `apply_values_to_parameters`: the PR had restructured the function in a way that broke `user_prompt` + `file` parameter handling (user-prompted file paths would be inserted raw instead of reading file content). The merge restored main's correct implementation. The deleted test `test_build_recipe_user_prompt_file_parameter_reads_file_content` was also restored.

All recipe tests pass. Clippy and fmt are clean.

Labeling `needs_human` for a maintainer to make the final call — the feature is well-motivated (discussion #8917) and the code is solid, but this is a first-time contributor with a feature PR.

**Comment by lifeizhou-ap** [comment]
Hi @jordigilh,

Thanks for your contribution! The problem this is trying to solve makes sense to me: automation/orchestrator callers often already have structured incident context, enrichment data, and prior findings, and flattening that into a long list of scalar recipe params is awkward.

 I'm wondering if we should take one step back before adding `object` / `array` as recipe parameter types.

 A few things I’m still thinking through:

  - `input_type: object` / `array` does not describe the expected nested shape by itself. That makes the caller contract ambiguous: callers have to infer required fields from the template or examples, and Goose can only validate the top-level type. A mismatch can pass validation and only fail during rendering, or produce a prompt with missing context.  
  - Adding schema would make the contract clearer, but it also makes recipes much heavier to author.
  - Structured parameter could make the implementation more complex across template rendering, validation, and sub-recipe parameter passing.
  - For large arrays like `prior_findings`, template loops can create very large repetitive prompts, so I’m not sure that is the abstraction we want to encourage.

For the incident/orchestrator use case, I wonder if a simpler pattern is to keep recipe params scalar and pass dynamic structured data by file path.

For example:

  ~~~yaml
  parameters:
    - key: incident_context_file
      input_type: string
      requirement: required
      description: Path to the incident context file generated by the orchestrator

  instructions: |
    The file at `{{ incident_context_file }}` contains incident context generated by the orchestrator.

    Read the file before starting. The file could be JSON and look like this:

    {
      "signal": {
        "name": "OOMKilled",
        "namespace": "production",
        "severity": "critical",
        "resource": {
          "kind": "Pod",
          "name": "api-server-abc123"
        }
      },
      "enrichment": {
        "owner_chain": "Pod/api-server-abc123 -> ReplicaSet/api-server -> Deployment/api-server"
      },
      "prior_findings": [
        {
          "name": "change-freeze-check",
          "phase": "pre-investigation",
          "output": "No active change freeze for namespace production"
        }
      ]
    }

    Use nested fields from `signal`, such as `signal.namespace`, `signal.severity`,
    `signal.resource.kind`, and `signal.resource.name`, to identify the affected resource.

    `prior_findings` is an array. For each finding, read its `name`, `phase`, and
    `output`, then use those values when deciding the next action.

    Use `enrichment.owner_chain`, if present, to understand ownership.

    Check the CMDB for the affected resource and recommend next actions.
  ~~~

The orchestrator can generate a file and pass only the file path. This is more flexible because the data becomes an artifact the agent can work with, instead of a fixed template-rendering input.

A few benefits:

  - recipe parameters stay simple
  - the structured input data can be JSON, YAML, CSV, Markdown, or another file format without changing the recipe parameter model
  - the agent is not limited to template-time access like `signal.namespace` or `{% for finding in prior_findings %}`; it can decide how to use the file during the run, such as reading it, searching it, summarizing it, comparing fields, or focusing only on the relevant parts

I do think structured params could still be useful for workflows where the recipe template itself needs to branch or loop over structured values. I’m wondering whether the file-path pattern might cover the main use case here while keeping recipe parameters simpler.

**Comment by jordigilh** [comment]
Hi @lifeizhou-ap,

Thanks for the thoughtful feedback — these are the right questions to ask before merging a feature like this.

I agree that `input_type: object`/`array` alone doesn't describe the expected nested shape, and that callers shouldn't have to reverse-engineer the contract from template expressions. I'd like to propose addressing that directly rather than moving to the file-path pattern.

**The proposal: add a `schema` field to structured parameters.**

For `input_type: object` and `array`, the parameter definition would include a JSON Schema (Draft 7) that describes the expected shape. The schema would be validated at recipe build time — before the data reaches the template renderer — so malformed input fails fast with a clear validation error instead of surfacing as a confusing MiniJinja failure or a silently broken prompt.

Two forms would be supported:

- **`schema`** (inline) — for small, recipe-specific schemas defined directly in the recipe YAML
- **`schema_file`** (file reference) — for shared schemas reused across multiple recipes, resolved relative to the recipe root

Exactly one of the two is required for `object`/`array` parameters. Scalar parameter types ignore both fields.

This isn't a novel pattern — it's how parameter validation already works across the ecosystem:

- **OpenAPI 3.1** — every parameter object has a `schema` field that takes a full JSON Schema to describe the data shape. Structurally identical to what we're proposing here.
- **Tekton Pipelines** — [TEP-0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) added `object` parameter types with JSON Schema `properties` to define the expected shape, explicitly choosing JSON Schema as the validation approach.
- **Helm** — `values.schema.json` validates the entire values structure against JSON Schema Draft 7 before template rendering.

Here's what a recipe would look like:

```yaml
parameters:
  - key: signal
    input_type: object
    requirement: required
    description: Incident signal from the orchestrator
    schema:
      type: object
      required: [name, namespace, severity]
      properties:
        name:
          type: string
        namespace:
          type: string
        severity:
          type: string
          enum: [critical, high, medium, low]
        resource:
          type: object
          properties:
            kind:
              type: string
            name:
              type: string
      additionalProperties: false

  - key: prior_findings
    input_type: array
    requirement: optional
    default: "[]"
    description: Results from earlier investigation phases
    schema_file: schemas/investigation_finding.schema.json
```

**How this addresses your concerns:**

1. **"Doesn't describe the nested shape"** — The schema is the explicit contract. Callers know exactly what fields are required, what types they are, and what values are valid. No inference from templates needed.

2. **"Adding schema makes recipes heavier to author"** — The schema is the same JSON Schema that Helm uses for `values.schema.json` and that Kubernetes uses for CRDs. The audience for Goose recipes (SREs, platform engineers) already knows this format. For shared schemas, `schema_file` means you write it once and reference it from every recipe that uses it.

3. **"Structured parameters increase complexity across rendering, validation, and sub-recipe passing"** — Validation is a single call to the `jsonschema` crate (62M+ downloads, Draft 7 support). The rendering and sub-recipe passing paths don't change — they already work with `serde_json::Value`. The schema adds a validation gate before the existing pipeline, not a new pipeline.

4. **"Large arrays create repetitive prompts"** — The schema can enforce `maxItems` at validation time, so oversized arrays are rejected before they ever reach the template.

**Why I think the file-path alternative is risky for the orchestrator use case:**

The file-path pattern delegates parsing and interpretation of structured data entirely to the model. The recipe instructions have to embed both the input schema (as natural language describing the file format) and the behavioral expectations for each field in the prompt. That means validation also shifts to the model — if a caller passes a malformed file (missing fields, wrong types, invalid structure), the model has to detect the problem, decide how to recover, and potentially hallucinate values for missing fields rather than failing cleanly. With schema-based validation, we catch structural errors *before rendering*, at the parameter validation boundary. The recipe either receives valid, fully-typed data or it doesn't run. There's no ambiguity and no reliance on the model's ability to gracefully recover from bad input.

This distinction matters most for deterministic orchestration — where `{{ signal.namespace }}` must resolve to a specific value, not a model's best guess at what field to read from a JSON file. Template rendering with pre-validated data is strictly more reliable than hoping the model correctly parses an arbitrary file and extracts the right values.

The file-path pattern is also already possible today with `input_type: string` and no code changes, so it doesn't need to be the *only* path. I think both patterns have value: file paths for unstructured/exploratory data, structured params with schema for deterministic template rendering. This PR covers the latter.

Happy to prototype the `schema`/`schema_file` fields if this direction makes sense.

**Comment by jordigilh** [comment]
@lifeizhou-ap any thoughts on the proposal I posted last week on how to address the schema validation on nested objects?

**Comment by lifeizhou-ap** [comment]
Hi @jordigilh, sorry for the late reply.

Thanks for putting that together. A few things on my mind:

- Would the file-path / inline-JSON pattern I shared earlier work for your case? If there's something it can't do, that'd help me understand.
-  Even with the jsonschema crate doing the validation, object/array params still touch a lot of other places. CLI/UI input, defaults, sub-recipe passing, template validation and rendering.  That's a fair bit of complexity for both authors and maintenance.
- For the orchestrator case, I'm wondering if letting the agent handle the structured input is enough. Recipes are getting closer to agent specs and we would like to keep the param model simple if we can.

Going to close this one for now. If more people in the community ask for structured params, happy to revisit.

**Comment by jordigilh** [comment]
Hi @lifeizhou-ap , thanks so much for the time and feedback. Sounds good to me to close this PR and wait for further feedback. 

Cheers,

/Jordi
